pennylane/templates/embeddings/qaoa.py

```diff
@@ -260,22 +200,136 @@ def circuit(weights, f=None):
-        qaoa_ising_hamiltonian(weights[l], wires, local_fields)
+    num_params = 2
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, features, weights, wires, local_field="Y", do_queue=True):
+
+        if local_field == "Z":
+            self.local_field = qml.RZ
+        elif local_field == "X":
+            self.local_field = qml.RX
+        elif local_field == "Y":
+            self.local_field = qml.RY
+        else:
+            raise ValueError(f"did not recognize local field {local_field}")
+
+        super().__init__(features, weights, wires=wires, do_queue=do_queue)
+        self._preprocess()
+
+    def expand(self):
+
+        features = self.data[0]
+        weights = self.data[1]
+
+        # first dimension of the weights tensor determines
+        # the number of layers
+        repeat = qml.math.shape(weights)[0]
+
+        with qml.tape.QuantumTape() as tape:
+
+            for l in range(repeat):
+                # apply alternating Hamiltonians
+                qaoa_feature_encoding_hamiltonian(features, self.wires)
+                qaoa_ising_hamiltonian(weights[l], self.wires, self.local_field)
+
+            # repeat the feature encoding once more at the end
+            qaoa_feature_encoding_hamiltonian(features, self.wires)
+
+        return tape
+
+    def _preprocess(self):
+        """Validate and pre-process inputs as follows:
+
+        * Check that the features tensor is one-dimensional.
+        * Check that the first dimension of the features tensor
+          has length :math:`n` or less, where :math:`n` is the number of qubits.
+        * Check that the shape of the weights tensor is correct for the number of qubits.
+        """
+
+        features = self.parameters[0]
+        weights = self.parameters[1]
+
+        shape = qml.math.shape(features)
+
+        if len(shape) != 1:
+            raise ValueError(f"Features must be a one-dimensional tensor; got shape {shape}.")
+
+        n_features = shape[0]
+        if n_features > len(self.wires):
+            raise ValueError(
+                f"Features must be of length {len(self.wires)} or less; got length {n_features}."
+            )
 
-    # repeat the feature encoding once more at the end
-    qaoa_feature_encoding_hamiltonian(features, wires)
+        shape = qml.math.shape(weights)
+        repeat = shape[0]
+
+        if len(self.wires) == 1:
+            if shape != (repeat, 1):
+                raise ValueError(f"Weights tensor must be of shape {(repeat, 1)}; got {shape}")
+
+        elif len(self.wires) == 2:
+            if shape != (repeat, 3):
+                raise ValueError(f"Weights tensor must be of shape {(repeat, 3)}; got {shape}")
+        else:
+            if shape != (repeat, 2 * len(self.wires)):
+                raise ValueError(
+                    f"Weights tensor must be of shape {(repeat, 2*len(self.wires))}; got {shape}"
+                )
+
+    @staticmethod
```

Comment:
I guess this was discussed in the ADR, but I'm not a huge fan of performing parameter initialization inside the operation itself. It's pretty straightforward for the user to just add a line with `np.random.normal()` or `np.random.unifrom()` to initialize parameters

Reply:
Yes, but have you ever tried to figure out what the correct dimensions for `CVNeuralNetLayers` are (just to name an example)? This takes me 5 min of digging through docstrings each time, and I wrote most of the templates :P

Reply:
So maybe consider this a utility function for people like me :)

Reply:
Same here! I use PennyLane daily, and still struggle with parameter initialization for a lot of templates (especially CV) (except `StronglyEntanglingLayers` 🙏 )

Reply:
I think that is a nice idea, thought about this as well...I think I'll use this